bob-backend/src/pearl/group.rs

```diff
@@ -170,54 +172,37 @@ impl Group {
     // create pearl for current write
     async fn create_write_pearl(
         &self,
-        ts: u64,
+        data: &BobData,
         timestamp_config: StartTimestampConfig,
     ) -> Result<(ChildId, Holder), Error> {
-        let mut indexes = self.created_holder_indexes.write().await;
-        let created_holder_index = indexes.get(&ts).copied();
-        let index = if let Some(exisiting_index) = created_holder_index {
-            exisiting_index
+        let holders = self.holders.upgradable_read().await;
+        
+        let created_holder_index = Self::find_actual_holder(&holders, data).await;
+        Ok(if let Ok(index_and_holder) = created_holder_index {
+            index_and_holder
         } else {
-            let new_index = self
+            info!("creating pearl for timestamp {}", data.meta().timestamp());
+            let pearl = self.create_pearl_by_timestamp(data.meta().timestamp(), &timestamp_config);
+            self
                 .settings
                 .config()
                 .try_multiple_times_async(
-                    || self.try_create_write_pearl(ts, &timestamp_config),
+                    || pearl.prepare_storage(),
                     "pearl init failed",
                     self.settings.config().settings().create_pearl_wait_delay(),
                 )
                 .await?;
-            debug!("group create write pearl holder index {:?}", new_index);
-            indexes.insert(ts, new_index);
-            debug!("group create write pearl holder inserted");
-            new_index
-        };
-        Ok((
-            index,
-            self.holders
-                .read()
-                .await
-                .get_child(index)
+            debug!("backend pearl group save pearl storage prepared");    
+            let mut holders = RwLockUpgradableReadGuard::upgrade(holders).await;
+            let new_index = holders.push(pearl).await;
+            debug!("group create write pearl holder inserted, index {:?}", new_index);
+            let holder = holders
```

Comment:
You've just added `pearl` and got its index. There is no reason to take it from the collection. You can just return `(new_index, pearl)`